Add user-level email address visibility setting.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This PR adds user-level email address visibility setting (and also the realm-level default for new users) and removes the existing realm setting.
Have explained major changes in details in the fourth and fifth commit which are major commits.
And we can squash some commits, but have kept it separate now for review.

Some important points -
- There is change in logic for how we set client_gravatar to false based on user setting, which means avatar_url field can be available for some users and not for others.
- Have not made any changes to typeahead and the behavior is same (including showing fake emails) except that user-level setting is used.
- We need to think how the sort and search should work in users/deactivated users list and subscribers list. In the users list for now the "ID: 16" is just considered a string like email and search and sort is done accordingly.
- I also wanted to confirm one thing, that we should always show user's own real email to themselves in all places like user popover, users list in settings, etc. irrespective of setting right?

Work left -
- Updating and adding help documentation.
- Prototype adding `?` besides `ID: 26` mentioning that email is hidden by the user.

 <!-- How have you tested? -->

 <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

Issue - #20035

sahil839 commented:
Also need to do proper manual testing once more.

Just attaching some screenshots with how email and ID look together in a column, in case if someone wants to give an early feedback - 
![Screenshot from 2021-10-30 00-46-14](https://user-images.githubusercontent.com/35494118/139490644-8d5cc244-ffe5-4d16-9f8e-195e32cd16ab.png)

![Screenshot from 2021-10-30 00-45-56](https://user-images.githubusercontent.com/35494118/139490652-7f11814d-2611-4e6b-8175-a3af263baa75.png)

timabbott commented:
Hmm, I'm not thrilled with how that column ends up looking.  I wonder if we want to go more in the direction of a "Copy user ID" button for the table?  It may be worth more detailed discussion in general -- e.g. do we want to show user IDs in typeahead?  @alya FYI.

timabbott commented:
I merged the first two prep commits as the series ending with 9a03a78e3d1e7fc18e48e49f6a51c93afaf739a5.

alya commented:
Hm, what do we think is the main purpose of showing IDs in the users list? I think it's more to confirm someone's identity than to copy-paste the ID somewhere?

In any case, discussing in #design sounds good.

timabbott commented:
If you want to split out more readily-merged preparatory commits, I think the `settings_data.show_email(user)` parameter addition (where the parameter doesn't do anything) would be such a refactoring.

timabbott commented:
> I think it's more to confirm someone's identity than to copy-paste the ID somewhere?

It's for both.  I think mainly to confirm identity, and secondarily it's useful if you're doing something with the API by hand.

alya commented:
OK. In that case, having a "copy" button too seems fine, but I wouldn't want to hide the ID from view.

sahil839 commented:
> If you want to split out more readily-merged preparatory commits, I think the `settings_data.show_email(user)` parameter addition (where the parameter doesn't do anything) would be such a refactoring.

This gives a lint error of unused variable. We can add an exception for this and then remove it in further commit, but not sure whether it is worth doing.

sahil839 commented:
Not ready for review yet.

sahil839 commented:
> > If you want to split out more readily-merged preparatory commits, I think the `settings_data.show_email(user)` parameter addition (where the parameter doesn't do anything) would be such a refactoring.
> 
> This gives a lint error of unused variable. We can add an exception for this and then remove it in further commit, but not sure whether it is worth doing.

This is ready for next review. The CodeQL errors are for the urls that existed before also, not sure what to do here.
Added a couple of small prep commits and also a commit to sort subscriber-list by name since some rows will have "hidden" and some will have real email. Do we still want to sort by email if realm-level default settings say email is visible and by name in other case? That would mean "(hidden)" will be grouped together.

Also, we still need to discuss what do we need to do for fake emails present in typeaheads and search box.

alya commented:
> Also, we still need to discuss what do we need to do for fake emails present in typeaheads and search box.

I *really* dislike showing fake emails in the search box. When I've commented on this in the past, @timabbott has pointed to a bigger search area rewrite (conversion to using pills) as how we'll deal with it. But I don't see that happening any time in the near future, so if there's a reasonable fix before then, I'd be very happy to see it. 

For example, does anything prevent us from showing what our @-mention auto-complete produces, i.e. either just the name or name|id when it's ambiguous?

alya commented:
What are the typeaheads where fake emails come up?

alya commented:
> Do we still want to sort by email if realm-level default settings say email is visible and by name in other case? That would mean "(hidden)" will be grouped together.

It seems like we currently just sort by name by default -- is that right? Defaulting to sorting by name seems reasonable...

sahil839 commented:
> > Do we still want to sort by email if realm-level default settings say email is visible and by name in other case? That would mean "(hidden)" will be grouped together.
> 
> It seems like we currently just sort by name by default -- is that right? Defaulting to sorting by name seems reasonable...

Currently we sort by emails if real emails are accessible and names otherwise.

sahil839 commented:
> What are the typeaheads where fake emails come up?

Sorry, I think I was confused and noticed this while testing this PR itself and thought this was the current behavior in main. We do not show fake emails in typeaheads except the search suggestions typeahead. Also, while checking this I noticed we show fake emails in the `People to add` section when creating new stream.

sahil839 commented:
> > Also, we still need to discuss what do we need to do for fake emails present in typeaheads and search box.
> 
> I _really_ dislike showing fake emails in the search box. When I've commented on this in the past, @timabbott has pointed to a bigger search area rewrite (conversion to using pills) as how we'll deal with it. But I don't see that happening any time in the near future, so if there's a reasonable fix before then, I'd be very happy to see it.
> 
> For example, does anything prevent us from showing what our @-mention auto-complete produces, i.e. either just the name or name|id when it's ambiguous?

I tried changing the search box to only show names if real emails are not visible around a year back, but was stuck somewhere which I do not remember now, will look into code today and report back here.

sahil839 commented:
I looked at search code, the emails in typeahead can be removed easily, but not sure what amount of work will be required to remove the fake email appearing in the search box on selecting one of the options from typeahead (for example with a `pm-with` operator).
From my understanding of code, it is not something trivial and would take time.

timabbott commented:
The server/client protocol also expects email addresses; changing that model is a large project.  I think changing just the typeahead to match how the compose box typeahead works is a pretty reasonable option, though.

alya commented:
> > > Do we still want to sort by email if realm-level default settings say email is visible and by name in other case? That would mean "(hidden)" will be grouped together.
> > 
> > 
> > It seems like we currently just sort by name by default -- is that right? Defaulting to sorting by name seems reasonable...
> 
> Currently we sort by emails if real emails are accessible as per email_address_visibility setting and names otherwise.

Ok, it seems simple and reasonable to just always sort by name?

alya commented:
> The server/client protocol also expects email addresses; changing that model is a large project. I think changing just the typeahead to match how the compose box typeahead works is a pretty reasonable option, though.

That would be great!

sahil839 commented:
> The server/client protocol also expects email addresses; changing that model is a large project. I think changing just the typeahead to match how the compose box typeahead works is a pretty reasonable option, though.

Do you mean typeaheads used for mentioning users in the compose box here? I do not understand what behavior is desired here.

sahil839 commented:
I looked at the search code bit more. The main problem is that we need to use `email` (which can be real or fake email depending upon setting) field of the person object for acutally getting the user ids since we only maintain a mapping of email field with the person objects currently. And we need to use email in the search code as our filter system does not work completely with ids as of now. And with the possibility of search string consisting of multiple parts, not sure how we would get it to work such that we display something other than email in the input but actually use the email for activating the narrow according to search. I will be more sure once I get to know what the actual behavior is desired.

And I think, the rest of the work is probably ready for review aside from the emails in search input.

alya commented:
This is what I was imagining for how the search changes from a UI perspective:

1. When you select a user from the search-box auto-complete (e.g. from:...), what get's filled in is something like `sender:"Tim Abbott"` (or `sender:"Tim|496"` when the name is not unique), rather than the fake email.
2. Entering part of a fake email does not trigger the typeahead for that email; fake emails would be completely invisible to users. E.g. this would no longer work:

<img width="201" alt="Screen Shot 2021-11-15 at 5 01 51 PM" src="https://user-images.githubusercontent.com/2090066/141876777-cd234a1d-a1e1-4fbc-a8cc-433f3e8d246d.png">

@timabbott is that what you had in mind?

sahil839 commented:
I tried it further to show names in the search box. We can do this by modifying the operand to contain emails instead of names just before passing it to `narrow.activate` which will involve two loops one for operators and one over all names (in case of searching for group-PMs. And will need some additional code to extract the id in case of duplicate names like we do for mention syntax. Will do a prep PR to get the idea of exactly how the code will look by tomorrow.

alya commented:
Great, thanks for digging into this!

timabbott commented:
What's the status of this?  I can commit some time to doing a review later this week but am not sure what known issues/problems remain.

sahil839 commented:
> What's the status of this? I can commit some time to doing a review later this week but am not sure what known issues/problems remain.

The known problem is about the search input. The search typeahead can be fixed to show real-emails when available, but changing the search input content which appears after selecting an option from typeahead to show names is non-trivial fix. I have tried doing it in #20282, but am not too confident about the approach. Would be great if you have a look at #20282.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5282,6 +5294,32 @@ def do_change_user_setting(
+            user_profile.realm,
+            dict(type="realm_user", op="update", person=person),
+            active_user_ids(user_profile.realm_id),
+        )
```
We generally never do a `send_event` call before changing the actual values in the database; so we should reorder this block to follow that convention.

sahil839 replied:
We are sending event for each field after updating it. Do we want to first update `email` and then send events for updating `email_address_visibility`, `email` and `avatar_url` at last?

timabbott replied:
All events should be only after the database update has been committed that it is reflecting to clients; it's I suppose OK to have update1-event1-update2-event2. and also we should use the `transaction.on_commit`  approach to make sure events are only sent if the transaction commits.  But the second part is not critical -- or at least a bug we have everywhere here.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5282,6 +5294,32 @@ def do_change_user_setting(
+
+        flush_user_profile(sender=UserProfile, instance=user_profile)
+
+        notify_avatar_url_change(user_profile)
```
I think there's a subtle issue here, where if a user changes their email address visibility to be more restrictive, we'd do an event that deletes the `delivery_email` values from others' clients' memory.  This may be a reason to send an `email_address_visibility` event even if we don't add those to user objects in general.

timabbott replied:
I suppose one answer would be to send a `update_user` event with `delivery_email=null` as the event format to the set of users who lost access?

timabbott commented on zerver/lib/bulk_create.py:
```diff
@@ -44,11 +53,15 @@ def bulk_create_users(
     else:
         for user_profile in profiles_to_create:
             user_profile.email = user_profile.delivery_email
+            user_profile.email_address_visibility = realm_user_default.email_address_visibility
```
It's kinda weird that we're both passing this value into `create_user_profile` and then applying it here.  I think the structural bug here is that `create_user_profile` ignores `RealmUserDefault` in general, and this bulk-creation code path ends up ignoring it as a result.  

I think the correct fix here would be to just do a prep commit to have the `bulk_create_users` code path do this:
```
    if bot_type is None:
        realm_user_default = RealmUserDefault.objects.get(realm=realm)
        copy_default_settings(realm_user_default, user_profile)
```

in the loop generating `profiles_to_create`, instead of adding this line.

timabbott commented on zerver/lib/create_user.py:
```diff
@@ -115,9 +116,13 @@ def create_user_profile(
+        bot_type
+        or realm_user_default.email_address_visibility
+        == RealmUserDefault.EMAIL_ADDRESS_VISIBILITY_EVERYONE
+    ):
```
Did we need to replace `user_profile.email_address_is_realm_public` with this more complex conditional?  We still use it in this file, and I think the value is unchanged from before.

Oh, I guess the issue is that we haven't set `user_profile.email_address_visibility yet`?  

Maybe it'd be a cleaner interface to just pass `email_address_visibility` into this function from the caller, rather than the whole `RealmUserDefault` object, set it in the `UserProfile` invocation (with a comment that we need to do this early for the stated reason), and then have this call `user_profile.email_address_is_realm_public` as before.

timabbott commented on zerver/lib/message.py:
```diff
@@ -329,6 +329,11 @@ def finalize_payload(
+        if obj["sender_email_address_visibility"] != UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
+            # If email address of the sender is only available to administrators,
+            # clients cannot compute gravatars, so we force-set it to false.
+            client_gravatar = False
```
We aren't really in a position to do plumbing for this, but I think the comment should perhaps have an added sentence "If we plumbed the current user's role, we could allow client_gravatar=True here if the current user's role has access to the target user's email address"

timabbott commented on zerver/lib/users.py:
```diff
@@ -546,10 +548,13 @@ def get_raw_user_data(
             custom_profile_field_data = profiles_by_user_id.get(row["id"], {})
-
+        if row["email_address_visibility"] != UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
+            client_gravatar = False
```
Mutation of inputs like this is messy.  A cleaner way to do this loop would be to just pass to `format_user_row` the value:

client_gravatar_for_user=client_gravatar and row["email_address_visibility"] == UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE

timabbott commented on zerver/tests/test_events.py:
```diff
@@ -2350,6 +2364,15 @@ def do_change_user_settings_test(self, setting_name: str) -> None:
+                ]:
+                    num_events = 4
+                else:
+                    num_events = 2
```
It's worth adding a comment explaining why the number of events varies for this sort of thing.

timabbott commented on zerver/views/message_fetch.py:
```diff
@@ -991,12 +991,7 @@ def get_messages_backend(
-        # If email addresses are only available to administrators,
-        # clients cannot compute gravatars, so we force-set it to false.
+    if is_web_public_query:
         client_gravatar = False
```
Do we want to add a comment here noting that `client_gravatar` here is just the user's requested value, and `finalize_payload` is responsible for sending the avatar URL or not depending on each individual sender's email_address_visibility?

timabbott commented:
This is great @sahil839!  In addition to the smaller comments above, I have the following notes:
* Why do we need to send the `email_address_visibility` values for other users in the `GET /users` API endpoints?  I would have thought that would be a personal setting that doesn't need to be shared with other users.  I guess the answer is that `user_can_access_delivery_email` requires it as written.  But I'm not sure it does -- can't we just check if the `person` object has a `delivery_email` field on it or not?  (BTW the comments in that function need updating for the new reality in any case).  I guess the function's comments say it is asking whether you should expect to have access to `delivery_email`; but I think it seems silly to sync around via the API a value that's redundant (all that matters for you is whether you can see the email, which you can check by `.delivery_email`).
* It looks like we lose the `help_settings_link` tag; I think it'd be helpful to add that link for the new setting.  (We'll also want to draft changes to the /help/ page, but that can happen as an immediate follow-up to merging this).
* This is too much complexity to consider in this PR, but we might be able to remove the computed `.email` field from `UserProfile` entirely after these changes, and just have it be a computed field on `UserProfile`.
* We probably should plan to add more defaults for other organization types beyond business/education in the future; and maybe also think about how we want this to work when doing data imports.  @alya FYI for this.
* The OpenAPI file should have `**Changes**` entries documenting the removed `email_address_visibility` field for the realm settings and register endpoints; usually we just put these in the global description for the endpoint, since it was deleted.  (In general, we should never have a change to `zulip.yaml` that doesn't come with a "Changes" entry elsewhere in `zulip.yaml`.  I think it can be brief, e.g. "**Changes**: The `email_address_visibility` setting was removed in Zulip 5.0 (feature level ...).  It was replaced by a [user setting] with a [realm user default], with the encoding of different values preserved. Clients can support all versions by supporting the current API and treating every user as having the realm's email address visibility value."

I think only the first item there potentially requires a substantial reworking of this implementation.

sahil839 commented:
> * Why do we need to send the `email_address_visibility` values for other users in the `GET /users` API endpoints?  I would have thought that would be a personal setting that doesn't need to be shared with other users.  I guess the answer is that `user_can_access_delivery_email` requires it as written.  But I'm not sure it does -- can't we just check if the `person` object has a `delivery_email` field on it or not?  (BTW the comments in that function need updating for the new reality in any case).  I guess the function's comments say it is asking whether you should expect to have access to `delivery_email`; but I think it seems silly to sync around via the API a value that's redundant (all that matters for you is whether you can see the email, which you can check by `.delivery_email`).

@timabbott There is no `delivery_email` field when the setting is set to `Everyone`. the `email` field is set to real email of the user. So, we cannot just assume always that if `delivery_email` field is absent then we want to hide the email in UI. So, we would need `email_address_visibility` value for checking when to show email in the UI.

sahil839 commented on zerver/lib/bulk_create.py:
```diff
@@ -40,8 +49,13 @@ def bulk_create_users(
+                if settings_name in ["default_language", "enable_login_emails"]:
+                    continue
+                value = getattr(realm_user_default, settings_name)
+                setattr(profile, settings_name, value)
```
Have not used `copy_default_settings` here because it calls `.save()` and we instead do `bulk_create`. Other way was to pass a variable and do an early return before `.save()` in `copy_default_settings` but I thought this would be cleaner.

timabbott replied:
OK, that's reasonable -- let's just add a comment explaining that ideally we'd use `copy_default_settings`, but we want to avoid the save that it does.

timabbott replied:
I adjusted this comment before merging, thanks!

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -3391,6 +3391,12 @@ paths:
+                                It was replaced by a [user setting](/api/update-settings) with a
+                                [realm user default](api/update-realm-user-settings-defaults), with the encoding of different
+                                values preserved. Clients can support all versions by supporting the current API and treating
+                                every user as having the realm's email address visibility value.
```
Though `update_dict` event was not sent on updating realm `email_address_visibility` setting, added this here just because the setting was documented here and also there was some recent discussion to use this event format.

timabbott replied:
Yes, this is correct given the patterns we have here now.

sahil839 commented:
Left some comments above (including the one for passing `email_address_visibility` to user objects).
Things left are - 
- Removing `delivery_email` field from the user objects on changing the setting as mentioned [here](https://github.com/zulip/zulip/pull/20102#discussion_r765425935). Do we also want to do reverse, i.e. eg- adding `delivery_email` field for admins when setting is changed from nobody to admins only?
- Adding help link widget.
- Fixing the user-deactivation modal text which contains email. Do we want to show user-id when email is hidden or just hide the email. Currenty the generated/fake email is shown. Can discuss more regarding design in #design probably.

timabbott commented:
> Fixing the user-deactivation modal text which contains email. Do we want to show user-id when email is hidden or just hide the email. Currenty the generated/fake email is shown. Can discuss more regarding design in #design probably.

I think ideally we'd show the user ID in this setting.  It can be a separate PR, since it's mostly independent of these changes.

timabbott commented:
> Removing delivery_email field from the user objects on changing the setting as mentioned here. Do we also want to do reverse, i.e. eg- adding delivery_email field for admins when setting is changed from nobody to admins only?

I think that's probably the right way to do it, but it's probably OK for the real-time sync to be temporarily such that you need to reload your browser to get access to emails for users who just shared theirs more broadly.  And I think it'd help manage the complexity of this PR to defer that work.

timabbott commented:
> @timabbott There is no delivery_email field when the setting is set to Everyone. the email field is set to real email of the user. So, we cannot just assume always that if delivery_email field is absent then we want to hide the email in UI. So, we would need email_address_visibility value for checking when to show email in the UI.

Hmm.  I think the right design is probably that we should remove this overloading strategy, and just send `delivery_email` to everyone if `EMAIL_ADDRESS_VISIBILITY_EVERYONE`. That way, you always have a fake email address, and additionally you have a real email address if and only if you have access to the users' email addresses (and can display that real email address when available).  

And then over time we'll aim to just delete the fake email addresses from the API entirely, but that's going to be a series of N chained migrations after we integrate this.

sahil839 commented on zerver/lib/actions.py:
```diff
@@ -4506,6 +4507,7 @@ def notify_avatar_url_change(user_profile: UserProfile) -> None:
         # Even clients using client_gravatar don't need the email,
         # since we're sending the URL anyway.
         user_id=user_profile.id,
+        email_address_visibility=user_profile.email_address_visibility,
```
Sending `email_address_visibility` here to compute `client_gravatar` in `apply_event`.

sahil839 commented on zerver/lib/actions.py:
```diff
@@ -583,6 +583,7 @@ def notify_created_user(user_profile: UserProfile) -> None:
         # later event.
         custom_profile_field_data={},
     )
+    person["email_address_visibility"] = user_row["email_address_visibility"]
```
Sending `email_address_visibility` here to compute `client_gravata`r in apply_event. It is removed in the user objects passed to clients as response of different endpoints.

timabbott replied:
OK, that's reasonable.

timabbott replied:
Actually, I'm not sure why `apply_event` would need access to this -- shouldn't the user get an event with `delivery_email` present or not depending on their role?

sahil839 replied:
We set `client_gravatar` to false only depending upon the `email_address_visibility` of user, irrespective of role. You also left a comment before about we cannot plumb the user role as of now.

sahil839 commented on zerver/lib/actions.py:
```diff
@@ -1388,6 +1389,38 @@ def send_user_email_update_event(user_profile: UserProfile) -> None:
+            user_profile.realm,
+            dict(type="realm_user", op="update", person=person),
+            delivery_email_not_visible_user_ids,
+        )
```
Implemented both parts - adding `delivery_email` field if restrictions are relaxed and removing it if restrictions are tightened. Skipping adding part led to some mismatching states in event tests.

Also used `new_delivery_email` parameter here such that it does not interfere with the event that updates `state["delivery_email"]` when user changes their email.

timabbott replied:
Why shouldn't this be identical to that event?  It seems like a client should do the same thing if it receives a new delivery email for another user either way.

sahil839 replied:
`state["delivery_email"]` is set to real email even if the user is member and has set their setting to `Admins only`. While for the same user, `delivery_email` of themselves in `raw_users` object is not available.

sahil839 replied:
I was actually thinking about this - suppose user is hamlet and is a member. Hamlet sets the setting to `admins only`. Then also the hamlet dict in `state["raw_user"]` available to hamlet does not have `delivery_email` field and due to this in many places like users list in settings, hamlet email will also be hidden to hamlet himself as we do not check separately for the self case and directly use email available with person object. This behavior was present before this PR as well.

Do we want this or should our own email address be visible at all places in UI irrespective of setting?

timabbott replied:
I think the right model is that even with EMAIL_ADDRESS_VISIBILITY_NOBODY, hamlet's own `delivery_email` should be present in his person dictionary?  I'll try to write a bit of a spec in a new comment below.

sahil839 commented on zerver/lib/event_schema.py:
```diff
@@ -1134,6 +1135,13 @@ def check_realm_update_dict(
             ("delivery_email", str),
         ],
     ),
+    delivery_email_for_other_users=DictType(
```
Any suggestion for this variable name?

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -507,7 +507,7 @@ paths:
                                         When the delivery email of a user changes.
 
-                                        Note: This event is only visible to admins.
+                                        Note: This event is only visible to the target user.
```
This event is sent to only the user whose email is changed as per the code.

sahil839 commented:
Note that there are still some places where fake emails are visible -
- Search input and typeahead suggestions. Though it is a simple change in typeahead suggestions, but I am not sure we would want to do that without input change because it can confuse user as typeahead would have different email and after selecting any of the options from typeahead the search box query would have different email.

timabbott commented:
Let me try to write a revised spec following thinking through these areas of friction. I think the cleanest/simplest way to make our security model around `delivery_email` have the following model:
* All API endpoints / data fetches are changed to send `delivery_email` for another user if and only if the email address visibility functions say you should.  Potentially we make the `delivery_email` field a required field that has the value `null` if you don't have permission, rather than omitting it when not available -- that seems simpler.
* When a user's email address visibility is changed, we send an event to all users in the realm who now have access with a new `delivery_email` value and possibly a new `email` value; and a second event to users who now don't have access a `delivery_email=null` change for that user ID.  I think this is readily done with 1-2 `send_event` calls in each of the functions that (1) change this setting, (2) create an account, or (3) change a user's role.  
* We do a preparatory commit to make `can_access_delivery_email` have a new signature: `can_access_delivery_email(acting_user, target_user)`.  To cover the corner case you noted, this function returns True if the two users are the same, even with the `EMAIL_ADDRESS_VISIBILITY_NOBODY` value, so that you can always see your own email address.
* The `apply_events` code path then doesn't need any extra data -- it should be able to just overwrite whatever you had for that user with the new `delivery_email` value, like you would for any other setting.  (It feels like a security risk to try to late-delete data from events that users shouldn't be able to get a copy of -- better to just never had it be in an event at all).
* Once we have the server sending `delivery_email` values consistently, we can migrate it to replace all the logic for determining whether it has a real email address / `delivery_email` with just `person.delivery_email != null`, which is clean and simple.
* I guess ultimately we just want the legacy `email` field to stop being modified and eventually be removed entirely.  The challenge with doing that is that we have a few endpoints -- most importantly search, but also a few API things like a way to get presence given someone's email address for internal apps -- which work with `.email` and except that to be sometimes a real email address. I think what we'd do is migrate those API endpoints to a new `access_user_by_delivery_email(acting_user, email_address, for_admin=False)`, which would return a UserProfile if (1) the email address corresponds to a user and (2) `can_access_delivery_email(acting_user, target_user)`. This seems like a better API than the current model where we do `get_user()` and that uses fake or real emails depending on how the organization is configured.  Perhaps to manage complexity, we can try doing a prep PR to introduce `can_access_delivery_email` and `access_user_by_delivery_email` and then replace existing callers of `get_user_including_cross_realm` and `get_active_user` with this new `access_user_by_delivery_email`?  I think that the API would produce identical results to the current model, without relying on lookups on the `UserProfile.email` field.
* It's possible that once we've done the above things and maybe a bit more frontend work, we can change the `.email` field to always be a fake email for all organizations, rather than making it's value conditional on the settings? 
* It's OK for this PR to still have fake emails appear in a few places, as long as it's not a major regression from `main`, and we have enough complexity with this migration.

Once we've done that series of changes to the internals, the hope is that the change to add the new settings would just change the settings and thus the implementation of `can_access_delivery_email(acting_user, target_user)`.  

Does that approach make sense?  I think it solves all the problems I'm aware of around the API, and the resulting API would feel pretty clean.

sahil839 commented:
@timabbott Updated the PR to change the handling of `delivery_email` fields as mentioned above. Kept the change of `always passing delivery_email field including case when visibility is set to "Everyone" and also as "None" if user does not have access` after the commit for changing backend code to use new setting because it makes code for sending events and its tests much easier and avoids us from adding some temporary checks for bot cases as well which would have been removed after migrating to use user-level setting.

Some things left (probably doing as followups will be better especially the frontend part)-
- Sending event to update `delivery_email` on changing role. This would involve calling `send_event` in a loop for every user as per our current events structure.
- Adding help link as suggested above and a couple of places where we can hide fake emails

sahil839 commented:
> @timabbott Updated the PR to change the handling of `delivery_email` fields as mentioned above. Kept the change of `always passing delivery_email field including case when visibility is set to "Everyone" and also as "None" if user does not have access` after the commit for changing backend code to use new setting because it makes code for sending events and its tests much easier and avoids us from adding some temporary checks for bot cases as well which would have been removed after migrating to use user-level setting.
> 
> Some things left (probably doing as followups will be better especially the frontend part)-
> 
> * Sending event to update `delivery_email` on changing role. This would involve calling `send_event` in a loop for every user as per our current events structure.
> * Adding help link as suggested above and a couple of places where we can hide fake emails

Added help link with the settings labels.

sahil839 commented:
> > @timabbott Updated the PR to change the handling of `delivery_email` fields as mentioned above. Kept the change of `always passing delivery_email field including case when visibility is set to "Everyone" and also as "None" if user does not have access` after the commit for changing backend code to use new setting because it makes code for sending events and its tests much easier and avoids us from adding some temporary checks for bot cases as well which would have been removed after migrating to use user-level setting.
> > Some things left (probably doing as followups will be better especially the frontend part)-
> > 
> > * Sending event to update `delivery_email` on changing role. This would involve calling `send_event` in a loop for every user as per our current events structure.
> > * Adding help link as suggested above and a couple of places where we can hide fake emails
> 
> Added help link with the settings labels.

@timabbott reminder for review when you are available.

sahil839 commented:
> > @timabbott Updated the PR to change the handling of `delivery_email` fields as mentioned above. Kept the change of `always passing delivery_email field including case when visibility is set to "Everyone" and also as "None" if user does not have access` after the commit for changing backend code to use new setting because it makes code for sending events and its tests much easier and avoids us from adding some temporary checks for bot cases as well which would have been removed after migrating to use user-level setting.
> > Some things left (probably doing as followups will be better especially the frontend part)-
> > 
> > * Sending event to update `delivery_email` on changing role. This would involve calling `send_event` in a loop for every user as per our current events structure.
> > * Adding help link as suggested above and a couple of places where we can hide fake emails
> 
> Added help link with the settings labels.

Rebased the PR. Also opened #20684 for fixing emails in user-deactivation modal.

sahil839 commented:
Last push was just to rebase the PR.

sahil839 commented:
> > @timabbott Updated the PR to change the handling of `delivery_email` fields as mentioned above. Kept the change of `always passing delivery_email field including case when visibility is set to "Everyone" and also as "None" if user does not have access` after the commit for changing backend code to use new setting because it makes code for sending events and its tests much easier and avoids us from adding some temporary checks for bot cases as well which would have been removed after migrating to use user-level setting.
> > Some things left (probably doing as followups will be better especially the frontend part)-
> > 
> > * Sending event to update `delivery_email` on changing role. This would involve calling `send_event` in a loop for every user as per our current events structure.
> > * Adding help link as suggested above and a couple of places where we can hide fake emails
> 
> Added help link with the settings labels.

Rebased the PR.

alya commented:
Playing around with the PR, I think we should also just hide the email in the user edit modal, rather than showing the fake email:
<img width="266" alt="Screen Shot 2022-02-28 at 9 35 43 AM" src="https://user-images.githubusercontent.com/2090066/156031680-72eb18c0-c1cc-46ec-a303-04b9ad8b70c3.png">

Otherwise looking good!

sahil839 commented:
Fixed the user edit modal. 

Now, the only place where fake emails are shown is the search box input after selecting an option from typeahead. Typeahead option don't show emails if a user does not have access to them.

showell commented:
Please ping me (@showell) once this gets merged, as it will likely require a one-line fix to my stream-create PR #21233.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -926,7 +942,44 @@ paths:
+                                            - 4 = nobody
+                                            - 5 = moderators only
+
+                                            **Changes**: New in Zulip 5.0 (feature level 119).
```
Added this and not using `components/schemas/EmailAddressVisibility` here because we need not need `replacing the realm-level setting` part in the changes entry here.

sahil839 commented:
Previous pushes were to just rebase the PR.

sahil839 commented:
> > Please ping me (@showell) once this gets merged, as it will likely require a one-line fix to my stream-create PR #21233.
> 
> My stream-create PR was merged last week or so, so no need to ping me. Instead, please just make sure any changes to stream-edit get appropriately fixed in the stream-create process, where relevant.

There should be no changes in the stream create UI, since we only changed the setting from realm-level to user-level except that now the user would be able to see their email address in the `Subscribers` list and typeahead irrespective of the email visibility setting.

sahil839 commented:
Rebased the PR.

timabbott commented:
@sahil839 do you want to try moving this to a new PR? GitHub's UI is pretty unwieldy with 200+ items, and I think we've probably resolved most of the things discussed above (and a new summary can be written at the top of the new PR).

timabbott commented on templates/zerver/api/changelog.md:
```diff
@@ -27,6 +27,9 @@ format used by the Zulip server that they are interacting with.
   `email_address_visibility` setting.
+* [`GET /events`](/api/get-events): Added user `email_address_visibility` field
+  to the person object in the events sent when adding a user and when user changes
+  their avatar.
```
Is this only in these events, or also in the `/register` response? Is it an internal implementation detail because we need it for `apply_events`, or do clients need it too?

sahil839 replied:
Yes, `email_address_field` is explicitly included in these events to decide whether we need to include `avatar_url` or client can compute themseleves.

timabbott commented on zerver/lib/create_user.py:
```diff
@@ -117,13 +118,14 @@ def create_user_profile(
-        # If emails are visible to everyone, we can set this here and save a DB query
-        user_profile.email = get_display_email_address(user_profile)
+        # If emails are visible to everyone, we can set this here to the original email and save a DB query
+        user_profile.email = email
```
I think `get_display_email_address` returned a value equal to `email` in this case, right? I think that might have been the motivation for calling the function here -- the small readability benefit of having us call the same function in either code path.

sahil839 replied:
Updated this to use `get_display_email_address`.

timabbott commented on zerver/lib/events.py:
```diff
@@ -738,6 +738,9 @@ def _draft_update_action(i: int) -> None:
         if event["op"] == "add":
             person = copy.deepcopy(person)
+            if person["email_address_visibility"] != UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
+                client_gravatar = False
```
Can't we check this one by checking whether `delivery_email` is None, without needing to provide the `email_address_visibility` field in the event?

timabbott commented on zerver/lib/users.py:
```diff
@@ -606,12 +606,16 @@ def get_raw_user_data(
             custom_profile_field_data = profiles_by_user_id.get(row["id"], {})
+        client_gravatar_for_user = (
+            client_gravatar
+            and row["email_address_visibility"] == UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE
```
Would it be better for this check to use `can_access_delivery_email(acting_user, ...)`?

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -431,6 +431,21 @@ paths:
+                                          type: integer
+                                          description: |
+                                            The [policy][permission-level] for which users in this organization can see
+                                            the real email address of the user.
```
Maybe "The policy this user has selected for which other users in the organization ...", just to help solidify that this is a personal setting.

sahil839 replied:
Updated this.

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -348,16 +348,6 @@ def test_admin_api_hide_emails(self) -> None:
-            with self.assertRaises(JsonableError):
-                do_events_register(user, get_client("website"), client_gravatar=True)
-            self.assertEqual(mock_request_event_queue.call_args_list[0][0][3], False)
-
```
Does whatever this was testing get tested somewhere else now?

sahil839 replied:
This was testing whether the `client_gravatar` value passed to `request_event_queue` is overridden. But now we do not change the `client_gravatar` value in `do_events_register` now and we instead check and set it when we actually want to decide whether to include user's avatar url or not.

timabbott replied:
OK, makes sense.

timabbott commented on zerver/views/users.py:
```diff
@@ -572,7 +571,10 @@ def get_user_data(
+    if (
+        target_user is not None
+        and target_user.email_address_visibility != UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE
+    ):
```
I think `get_raw_user_data` has a copy of this logic already; perhaps we can do a prep commit deleting this block?  https://github.com/zulip/zulip/pull/20102/commits/254b578f35b65a1acc811174252e1708aa05e691#r865306423

sahil839 replied:
Removed this in the same commit only since the code in `get_raw_user_data` is added that commit itself.

timabbott commented on zilencer/management/commands/populate_db.py:
```diff
@@ -332,6 +332,9 @@ def handle(self, *args: Any, **options: Any) -> None:
             realm_user_default.enter_sends = True
+            realm_user_default.email_address_visibility = (
+                RealmUserDefault.EMAIL_ADDRESS_VISIBILITY_ADMINS
+            )
```
Why does this need parens -- is it just for better line-wrapping?

sahil839 replied:
Parens got added by black.

timabbott commented:
OK I think that completes my comments on this PR. I'm really excited to see it this close; it's possible making a new PR isn't worth it -- up to you.
* Could `users: Always pass "delivery_email" field in user objects` be moved to the start of the branch easily? As noted on the item linked from my next bullet, this might be a useful tool in letting us remove a messy API quirk that gives me pause about merging this.
* I think the principle thing that is messy is adding the `email_address_visibility` to various update events -- I'd like to avoid that sort of pollution of our API if we can. One option is to avoid that is tricks like https://github.com/zulip/zulip/pull/20102#pullrequestreview-962516012; another is to just have `apply_events` break our usual convention and do a quick database query for a rare event type like avatar changes.

sahil839 commented on zerver/actions/create_realm.py:
```diff
@@ -188,7 +187,19 @@ def do_create_realm(
+            # Email address of users should be initially visible to admins only.
+            realm_default_email_address_visibility = (
+                RealmUserDefault.EMAIL_ADDRESS_VISIBILITY_ADMINS
+            )
```
This is not a new feature. Previously we set the realm `email_address_visibility` setting for the education type orgs to be `Admins only`. Here as we will remove the realm-level setting, we set the realm-level defaut to be `Admins only` so that users when joining will have that value.

timabbott replied:
Ahh, OK, that makes sense -- I was just looking in the wrong place for it being removed.

sahil839 commented:
@timabbott Updated the PR with following changes -
- Removed the duplicate code to set `client_gravatar`.
- Use `get_display_email_address` directly in `create_user_profile`.
- Updated the API documentation as suggested.

Regarding `client_gravatar` calculation at various places -
- We previously considered `client_gravatar` as `False` when email is not visible to everyone even when admins can actually access the real email. Not sure why we do this. So, I have made current changes as per that only. If we want clients to compute the avatar url if the user is admin and can see email of the user, then we can check whether `delivery_email` is `None`, but that needs to be done after the `users: Always pass "delivery_email" field in user objects` commit regarding which I will comment below. But in that case too, we would need to pass `delivery_email` field in the object send with avatar change event and then we can do as asked in [this comment](https://github.com/zulip/zulip/pull/20102#discussion_r865306423).

Regarding moving `users: Always pass "delivery_email" field in user objects` commit at start -
- I tried this before, but we need to make many changes to tests especially for bots for that, which then would be reverted on adding the user-level setting as mentioned [here](https://github.com/zulip/zulip/pull/20102#issuecomment-993302691) previously.

timabbott commented:
OK, makes sense regarding reordering commits being more work than it's worth. I think the main thing remaining is this bundle of comments, basically whether we can avoid adding `email_address_visibility` to the API. I'd prefer for privacy reasons to have the specific value of another user's email address visibility not be available to other users -- structurally they should just know if they have access or not.

* https://github.com/zulip/zulip/pull/20102#discussion_r865288381
* https://github.com/zulip/zulip/pull/20102#discussion_r865300096
* The second bullet in https://github.com/zulip/zulip/pull/20102#issuecomment-1117863990, as an alternative to including extra data in the event by just doing a quick database query in `apply_events` if needed.

sahil839 commented:
How about adding a separate commit to directly check whether `delivery_email` is `None` or not, after the `users: Always pass "delivery_email" field in user objects` commit. This will also mean that admins will receive the url if user has set visibility to `admins only` unlike previous cases where we would sent url only if email was visible to everyone.

timabbott commented:
Yeah, that plan sounds good.

sahil839 commented:
@timabbott I added a commit to always include `delivery_email` for bots in the user object sent to clients, since `email` field is already set to real email.

> How about adding a separate commit to directly check whether `delivery_email` is `None` or not, after the users: Always pass "delivery_email" field in user objects commit. This will also mean that admins will receive the url if user has set visibility to admins only unlike previous cases where we would sent url only if email was visible to everyone.

And regarding the changes I proposed above, changing the behavior to not send url when email is available to the user will mean we would not send the avatar url to an admin when visibility is set to `Admins only`. But [the comment here](https://github.com/zulip/zulip/blame/aa5536fe3046d76813015dd7db83297a213ca03a/zerver/tests/test_users.py#L361) says that some clients like mobile apps will not be able to compute url in that case. So, just wanted to confirm whether we need to move forward with this change or we first need some changes the mobile code.

timabbott commented:
@sahil839 I think you should close this PR and make a new one; the GitHub UI is unusable with how many events this one has. That's probably also a good opportunity to summarize where this is and any remaining things to resolve.

sahil839 commented:
Closing this and will open a new PR.
